src/learner.cc

```diff
@@ -193,60 +190,129 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
+        metric_names.emplace_back(value);
+      }
+    } else {
+      cfg[key] = value;
     }
   }
-  ~LearnerImpl() override {
-    auto local_map = XGBAPIThreadLocalStore::Get();
-    if (local_map->find(this) != local_map->cend()) {
-      local_map->erase(this);
+  // Short hand for setting multiple parameters
+  void SetParams(std::vector<std::pair<std::string, std::string>> const& args) {
+    for (auto const& kv : args) {
+      this->SetParam(kv.first, kv.second);
     }
   }
-  // Configuration before data is known.
-  void Configure() override {
-    if (!this->need_configuration_) { return; }
 
-    monitor_.Start("Configure");
-    auto old_tparam = tparam_;
-    Args args = {cfg_.cbegin(), cfg_.cend()};
+  void SetAttr(const std::string& key, const std::string& value) {
+    attributes[key] = value;
+    mparam.contain_extra_attrs = 1;
+  }
+
+  bool GetAttr(const std::string& key, std::string* out) const {
+    auto it = attributes.find(key);
+    if (it == attributes.end()) return false;
+    *out = it->second;
+    return true;
+  }
+
+  bool DelAttr(const std::string& key) {
+    auto it = attributes.find(key);
+    if (it == attributes.end()) { return false; }
+    attributes.erase(it);
+    return true;
+  }
+
+  std::vector<std::string> GetAttrNames() const {
+    std::vector<std::string> out;
+    for (auto const& kv : attributes) {
+      out.emplace_back(kv.first);
+    }
+    return out;
+  }
+
+  const std::map<std::string, std::string>& GetConfigurationArguments() const {
+    return cfg;
+  }
+};
+
+std::string const LearnerAttributes::kEvalMetric {"eval_metric"};  // NOLINT
 
-    tparam_.UpdateAllowUnknown(args);
-    auto mparam_backup = mparam_;
-    mparam_.UpdateAllowUnknown(args);
-    generic_parameters_.UpdateAllowUnknown(args);
-    generic_parameters_.CheckDeprecated();
+
+/* \brief A class doing all the configuration for Learner. */
+class LearnerConfiguration {
+ public:
+  void Configure(LearnerAttributes* attr) {
+    if (!attr->need_configuration) { return; }
+
+    attr->monitor.Start("Configure");
+    auto old_tparam = attr->tparam;
+    Args args = {attr->cfg.cbegin(), attr->cfg.cend()};
+
+    attr->tparam.UpdateAllowUnknown(args);
+    auto mparam_backup = attr->mparam;
+    attr->mparam.UpdateAllowUnknown(args);
+    attr->generic_parameters.UpdateAllowUnknown(args);
+    attr->generic_parameters.CheckDeprecated();
 
     ConsoleLogger::Configure(args);
-    if (generic_parameters_.nthread != 0) {
-      omp_set_num_threads(generic_parameters_.nthread);
+    if (attr->generic_parameters.nthread != 0) {
+      omp_set_num_threads(attr->generic_parameters.nthread);
     }
 
     // add additional parameters
     // These are cosntraints that need to be satisfied.
-    if (tparam_.dsplit == DataSplitMode::kAuto && rabit::IsDistributed()) {
-      tparam_.dsplit = DataSplitMode::kRow;
+    if (attr->tparam.dsplit == DataSplitMode::kAuto && rabit::IsDistributed()) {
```

Comment:
Are we able to remove data split mode in the near future?

Reply:
Depends.  Do we still want to support different split mode in the future?